### ISSUE TITLE
Don't Stop Wrapper.py Upon Control+Z

### ISSUE THREAD

**Issue by benbaptist** [benbaptist/minecraft-wrapper#521]
This should NEVER kill a program. My server ran out of SSD space, and I was just quickly hitting Control+Z to check and clear something when I suddenly killed the Minecraft server and Wrapper.py instance. This sucked, because this just resulted in it trying to save and flush when it's impossible to with literally 0 bytes of space available. Now the world is massively corrupted, a ton of wrapper's database files are corrupt, and Wrapper won't even start (that's a separate issue - it needs to handle these corrupted databases properly).

Even if it breaks curses or some other issue, it's pretty bad for a program to just stop like this when you expect different behavior like this. Because of this unexpected behavior, an action that normally wouldn't have ended so poorly ended up in chaos. I consider this a bug, even though technically it's probably deliberate, intended behavior.

**Comment by suresttexas00** [comment]
Ctrl-Z should not have caused all that.  It should have done the same as control-C, halting and saving the world and storages... But perhaps we simply have Z/C doing the opposite things.  I'll look into it this weekend.

**Comment by suresttexas00** [comment]
Ok, revisiting this now that I am back home.  Ctrl-Z is the linux code for SIGTSTP.  So this behavior should not have been unexpected.  We could have chosen to freeze the server instead, which is what a SIGTSTP properly does (https://stackoverflow.com/questions/11886812/whats-the-difference-between-sigstop-and-sigtstp).  The intent is to perform a controlled shutdown; which it attempted to do in your case.  However, given your lack of drive space, that was not possible.

I'm not sure why you were using control-Z, but you should have probably issued a /freeze of the server instead and attempt to make some space on the drive instead.  I don't think pressing any control key under those circumstances would have ended well.

So, I am not really calling this a bug.  Perhaps a feature request? 

Do you have any more information about the file corruption?  I have never run out of disk space before with wrapper, so I am curious.  I know we cannot control what the server.jar does under that circumstance.  However, I am curious:
1) Which files prevented wrapper start-up?
2) Were they json format or pkl (pickled) format?
3) Were the corrupted files "blank", or did they contain content?  I ask because I have noted instances where wrapper-properties.json file will get blanked out and prevent start-up (simply deleting the file and letting wrapper re-create it usually works). 

our shutdown code is in `core/wrapper.py`:

```

         handle_server()
        
        # upon ending the server handle, this all happens...
        self.plugins.disableplugins()
        self.log.info("Plugins disabled")
        self.wrapper_storage.close()
        self.wrapper_permissions.close()
        self.wrapper_usercache.close()
        self.log.info("Wrapper Storages closed and saved.")

        # use non-daemon thread to ensure alert gets sent before wrapper closes
        self.alerts.ui_process_alerts("Wrapper.py stopped at %s" % time.time(), blocking=True)

        # wrapper execution ends here.  handle_server ends when
        # wrapper.halt.halt is True.
        if self.sig_int:
            self.log.info("Ending threads, please wait...")
            time.sleep(5)
            os.system("reset")

    def signals(self):
        signal.signal(signal.SIGINT, self.sigint)
        signal.signal(signal.SIGTERM, self.sigterm)
        # noinspection PyBroadException
        try:
            # lacking in Windows
            signal.signal(signal.SIGTSTP, self.sigtstp)
        except:
            pass

    def sigint(*args):
        self = args[0]  # We are only interested in the self component
        self.log.info("Wrapper.py received SIGINT; halting...\n")
        self.sig_int = True
        self._halt()

    def sigterm(*args):
        self = args[0]  # We are only interested in the self component
        self.log.info("Wrapper.py received SIGTERM; halting...\n")
        self._halt()

    def sigtstp(*args):
        # We are only interested in the 'self' component
        self = args[0]
        self.log.info("Wrapper.py received SIGTSTP; NO sleep support!"
                      " Wrapper halting...\n")
        # this continues a frozen server, if it was frozen...
        os.system("kill -CONT %d" % self.javaserver.proc.pid)
        self._halt()

    def _halt(self):
        self.javaserver.stop(self.halt_message, restart_the_server=False)
        self.halt.halt = True

    def shutdown(self):
        self._halt()

```

**Comment by benbaptist** [comment]
Reason I Control+Z'd was to jump back to the console, not necessarily to freeze that server. Of course, normally I'd just switch to a different shell (different `screen` window or another Terminal tab) but habitually, I sometimes just hit Control+Z. As that article states, it should just freeze it so there shouldn't be any reason to stop the whole wrapper & server. SIGSTOP and SIGTSTP are generally-speaking for temporarily freezing a process (especially SIGSTOP), not to kill one. We already have Control+C for this, and just about literally every other piece of software on Linux would do exactly what I expected - a simple freeze, and return to the console. We really don't need to catch it at all. Just let the shell do its natural thing and we'll be good. 

And sure! Before resetting my Wrapper instance, I backed up the wrapper-data because I knew it would be useful to debug so we can handle future problems. And indeed, the server.jar stuff is all out of our control so no worries there. So I'll put the files somewhere.

**Comment by benbaptist** [comment]
^Made some small edits to my post above so if you read from emails, you might want to click on Github.

**Comment by suresttexas00** [comment]
I never read the email strings anymore, so no worries, I will always be reading it here on Github.

Well, I'm still mystified about ctrl-z being used to switch windows.  I use Ubuntu type Linuxes and Alt-Tab would be my "goto" keys to switch windows.  Ctrl-Z does not seem to do anything on my system (except undo in certain GUIs).

The difference here between Ctr-z and Ctrl-c is a subtle one here... Control-Z will unfreeze a frozen server so that it can save it's data. 

I must do some re-testing on this.  Our problem is that we have "two" applications running; the server jar and the python wrapper.  If one stops, does the other stop?  What are the consequences of stopping wrapper and letting the server continue?  And V.v.?  I think you are correct about the SIGSTOP/SIGTSTP applications, I just want to be sure nothing weird occurs if I de-implement the shutdown facilities.

**Comment by suresttexas00** [comment]
Lets say we implement ctrl-Z as a proper sigtstop and pause wrapper and server... should this be toggle-able?  ctrl-Z freezze, ctrl-Z resume..?

**Comment by suresttexas00** [comment]
Unfortunately, doing things by habit will always bite you in the rear.  Times without number have I halted my wrapper instances while attempting to "Ctrl-C" (copy) some text out of the wrapper window (versus the Ctrl-Alt-C that the terminal window uses for copy).
Had I been in the condition of no disk space at any of those times, I would have suffered the same "disaster" that you did...

**Comment by suresttexas00** [comment]
I found the reason I started lumping ctrl-Z in with ctrl-c:
https://github.com/benbaptist/minecraft-wrapper/issues/406

There was this weird wrapper bug that used to mess up the console after running wrapper.  I believe the bug was from unhandled Ctrl-Z's, but maybe not.

way down on the bottom of the discussion:
```
l33tlinuxh4x0r commented on Jan 9, 2017
weird
 @suresttexas00
     Collaborator
suresttexas00 commented on Jan 9, 2017 •  edited 
I googled this up and it looks like I should be supporting a SIGTSTP (ctrl-z) 
as well as SIGTERM AND SIGINT .. but it appears you are correct about the 
way they should work.
 @suresttexas00
     Collaborator
suresttexas00 commented on Jan 9, 2017
I thing the SIGTSTP should be implemented as being the same as the
 server /freeze command. pausing wrapper execution is just not a good idea.
 @suresttexas00
     Collaborator
suresttexas00 commented on Jan 9, 2017 •  edited 
I am guessing both wrapper and the server are receiving the SIGTSTP... meaning
 the server will hang in the background, messing up wrappers proc() into the server. 
I will re-investigate this whole schema we use for wrapper signal handling...
```

Probably best to manually implement a freeze of server and wrapper...

**Comment by suresttexas00** [comment]
So, another question; currently we don't do anything with players on a server freeze.  Should we kick them during a freeze?  I am thinking especially of proxy players...

**Comment by suresttexas00** [comment]
After fiddling with this for a couple days, I have discovered the original reason I trapped this for shutdown:

*It shuts you down anyway* (on many systems) and leaves the server orphaned:
```
[16:45:52] [Server thread/INFO]: Starting Minecraft server on *:25542
...
^Z
[1]+  Stopped                 ./start.sh
surest@surest-700-047c ~/Desktop/testservers/server $ 

```
that [1]+ stopped job is the server. 

I have not been able to successfully trap this and "freeze" the server while allowing wrapper to continue running, once I pull the 1.0.0 rc, you are welcome to create an experimental branch from development and experiment with it for yourself.

That RC is waiting on a few more things before I pull it to your account:
implement bug fixes and improvements:
- player.message broken with regards to '&' usages.  I broke this a several versions ago.
- whitelist needs a long overdue overhaul: #314
    - jspanos71 @jspanos71 Feb 04 17:28
     I'm having an issue when i try to turn on proxy mode for a whitelist server... the
     existing whitelist entries don't register as valid and doing /whitelist add at the
     console doesn't work either... is there something i'm missing? running the latest
     beta build within a couple of minor builds
- verify units and update docs if needed for adjustBackupInterval(self, desired_interval) (may have units of minutes now, not seconds...)
- When name changes occur, add option to convert local player name by changing the name and the offline filenames in serverFolder/world/playerdata/<uuid>.dat and serverFolder/world/stats/<uuid>.json
- Also re-implement the server conversion option to convert online player files to offline (and vs-vs?)

**Comment by benbaptist** [comment]
Sorry about not responding sooner - I read your messages but just didn't actually get around to typing a response yet.

>Well, I'm still mystified about ctrl-z being used to switch windows. I use Ubuntu type Linuxes and Alt-Tab would be my "goto" keys to switch windows. Ctrl-Z does not seem to do anything on my system (except undo in certain GUIs).

I think you might've misread what I wrote a tad - I was referring to how ctrl+z typically "suspends" the current process, and drops you to the same shell that you were currently in. I separately was mentioning switching Terminal tabs or `screen` windows as another alternative to that, but that wasn't really pertaining to the Ctrl+Z. Apologies for the confusion there.

Generally speaking, the only time a `SIGTSTP` should ever need to be caught is during an extremely realtime-sensitive application, that absolutely cannot continue properly if frozen. I would not consider a Minecraft Server even remotely close to this type of situation. It'll have no problem if frozen and resumed, other than the time-per-tick limit being hit and resulting in the server restarting itself. 

An application that is realtime-critical, for example, would be my Raspberry Pi's FM transmitting program (PiFmAdv). This must push samples in real-time to the GPIO pins and cannot be interrupted. The slightest interruption kills the ability to create a perfect waveform and will mess the transmission up. At that point, it must be killed and restarted to fix.

I was going to mention how I recalled the possibility that this was needed due to a curses-related thing and the console. if it's otherwise not needed (and our curses code should probably be fixed anyways to handle this type of situation anyways, since many other programs like `nano`, for example, properly suspend when you hit ctrl+z and resume properly), that we should just not catch  SIGTSTP` at all. Just let it do its thing.

And as for child processes like the server, I'm fairly positive those, by nature of how child processes work on Linux, will also be suspended accordingly. 

Oh dang. Just as I'm writing this, you posted anther post. Time to read!

**Comment by benbaptist** [comment]
>It shuts you down anyway (on many systems) and leaves the server orphaned:

That's strange. What systems are those? Considering how long I've been out of loop with the codebase, I'm assuming the way we launch processes have changed. Are we still using subprocess? Did we change that to something else? 

In the past, I can guarantee ctrl+z did literally no harm to Wrapper itself and the server child process. So it's interesting that it does that. 

Another idea: we could just do nothing when Ctrl+Z is hit, or at the very least, let the user **confirm** their operation before shutting down. Might even be a nice thing to have for a traditional keyboard interrupt as well. I'm sure there's times where people will accidentally hit ctrl+c and piss themselves off. 😛

**Comment by benbaptist** [comment]
Wait, I just re-read your post. This might be a goofy question, but did you actually do `fg` afterwards? I know the naming scheme here is super misleading, and this is why signals' names like `SIGTSTP` and `SIGSTOP` sound like they should KILL the process, but it's really just about suspending them. When it says stopped jobs, it's referring to "suspended" processes. Hit `fg` and it'll pull everything back up. As I mentioned before, if you have your tick setting in `server.properties` set to the default value (about 60 seconds) or something even lower, and you remain frozen for the time it shows, it'll result in the server rebooting after you resume (but not during).

**Comment by suresttexas00** [comment]
The problem is (linux mint 3.1 - ubuntu offshoot), I am being dropped all the way out of the wrapper interface.  Whatever else I type at that point is sent to bash (see how I am at the command line `surest@surest-700-047c ~/Desktop/testservers/server $`).  Meanwhile, whatever is left of wrapper and the server are "out in space", unreachable by my commands.  I can't even reach them via the minecraft client because they are "frozen" and I can't log into either the proxy or the server port.

**Comment by suresttexas00** [comment]
I remember this same thing used to happen way back before we implemented these signal traps.

**Comment by suresttexas00** [comment]
We use readline, not curses.  We do have a "curses/jline-like readkey functionality that is off by default.  Unfortunately, we can "trap" these signals, but they DO interrupt the current progrma flow.  Therefore, offering the user an option would boiler down the one thing: do what you have to do before wrapper dies!  That is because, say after ctrl-z; you can't make the program resume from wherever it was interrupted (9/10 it was waiting for keyboard input at the wrapper/server console).

**Comment by suresttexas00** [comment]
The only luck I have ever had is with the current trap setup.  I has always worked properly with no errors.  If you accidentally use a bad shortcut key (can't do anything about Ctrl-S/Ctrl-Q of course because SIGSTOP cannot be trapped), wrapper takes the best shot it can at a controlled shutdown that will save the server , disable the plugins,and save the storages.

**Comment by suresttexas00** [comment]
Here is where I am going with this:  "Misc" to the rescue!

Adding the following item:
```
"trap-ctrl-z": true,
```
It will default to True since that is the safest option.  Those who feel they can safely use ctrl-z with no adverse effects (like orphaned servers such as happens to me on my Mint 18.3 Cinnamon system) can set it to False.

This change will be in the next dev pull, hopefull in the next few days.
